src/main/java/org/reaktivity/nukleus/maven/plugin/internal/generate/EnumTypeGenerator.java

```diff
@@ -71,48 +93,64 @@ private NameConstantGenerator(
         }
 
         public NameConstantGenerator addValue(
-            String name)
+            String name, Integer value)
         {
-            builder.addEnumConstant(name);
+            if (value == null)
+            {
+                builder.addEnumConstant(name);
+            }
+            else
+            {
+                builder.addEnumConstant(name, TypeSpec.anonymousClassBuilder("$L", value).build());
+            }
             return this;
         }
     }
 
     private final class ValueOfMethodGenerator extends MethodSpecGenerator
     {
         private final List<String> values = new LinkedList<>();
+        private final Map<String, Integer> valueAssociatedWithConstant = new HashMap<>();
 
         private ValueOfMethodGenerator(
             ClassName enumName)
         {
             super(methodBuilder("valueOf")
                     .addModifiers(PUBLIC, STATIC)
-                    .addParameter(int.class, "ordinal")
                     .returns(enumName));
         }
 
         public ValueOfMethodGenerator addValue(
-            String name)
+            String name, Integer value)
         {
             values.add(name);
+            if (value != null)
+            {
+                valueAssociatedWithConstant.put(name, value);
+            }
             return this;
         }
 
         @Override
         public MethodSpec generate()
         {
-            builder.beginControlFlow("switch ($L)", "ordinal");
+            builder.addParameter(int.class, valueType == null ? "ordinal" : "value");
```

Comment:
```
final String discriminant = valueType == null ? "ordinal" : "value";
```
Then use `discriminant` instead of repeating `valueType == null` condition.